internal/api/rbac/roles.go

```diff
@@ -0,0 +1,842 @@
+func RoleToResources(
+	kargoRole *svcv1alpha1.Role,
+) (*corev1.ServiceAccount, *rbacv1.Role, *rbacv1.RoleBinding, error) {
+	sa := buildNewServiceAccount(kargoRole.Project, kargoRole.Name)
+	amendClaimAnnotation(sa, kargoapi.AnnotationKeyOIDCSubjects, kargoRole.Subs)
+	amendClaimAnnotation(sa, kargoapi.AnnotationKeyOIDCEmails, kargoRole.Emails)
+	amendClaimAnnotation(sa, kargoapi.AnnotationKeyOIDCGroups, kargoRole.Groups)
+
+	role := buildNewRole(kargoRole.Project, kargoRole.Name)
+	rules := make([]rbacv1.PolicyRule, len(kargoRole.Rules))
+	for i, rule := range kargoRole.Rules {
+		rules[i] = *rule
+	}
+	var err error
+	if role.Rules, err = NormalizePolicyRules(rules); err != nil {
+		return nil, nil, nil, fmt.Errorf("error normalizing RBAC policy rules: %w", err)
+	}
+
+	rb := buildNewRoleBinding(kargoRole.Project, kargoRole.Name)
+
+	return sa, role, rb, nil
+}
+
+func amendClaimAnnotation(sa *corev1.ServiceAccount, key string, values []string) {
+	existing := sa.Annotations[key]
+	if existing != "" {
+		values = append(strings.Split(existing, ","), values...)
+	}
+	values = dedupeStringSlice(values)
+	if len(values) == 0 {
+		return
+	}
+	sort.Strings(values)
+	if sa.Annotations == nil {
+		sa.Annotations = map[string]string{}
+	}
+	sa.Annotations[key] = strings.Join(values, ",")
+}
+
+func dropFromClaimAnnotation(sa *corev1.ServiceAccount, key string, values []string) {
+	values = dedupeStringSlice(values)
+	if len(values) == 0 {
+		return
+	}
+	values = removeFromStringSlice(strings.Split(sa.Annotations[key], ","), values)
+	if len(values) == 0 {
+		delete(sa.Annotations, key)
+		return
+	}
+	sort.Strings(values)
+	if sa.Annotations == nil {
+		sa.Annotations = map[string]string{}
+	}
+	sa.Annotations[key] = strings.Join(values, ",")
+}
+
+func buildNewServiceAccount(namespace, name string) *corev1.ServiceAccount {
+	return &corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      name,
+			Annotations: map[string]string{
+				kargoapi.AnnotationKeyManaged: kargoapi.AnnotationValueTrue,
+			},
+		},
+	}
+}
+
+func buildNewRole(namespace, name string) *rbacv1.Role {
+	return &rbacv1.Role{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      name,
+			Annotations: map[string]string{
+				kargoapi.AnnotationKeyManaged: kargoapi.AnnotationValueTrue,
+			},
+		},
+	}
+}
+
+func buildNewRoleBinding(namespace, name string) *rbacv1.RoleBinding {
+	return &rbacv1.RoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      name,
+			Annotations: map[string]string{
+				kargoapi.AnnotationKeyManaged: kargoapi.AnnotationValueTrue,
+			},
+		},
+		Subjects: []rbacv1.Subject{
+			{
+				Kind:      rbacv1.ServiceAccountKind,
+				Namespace: namespace,
+				Name:      name,
+			},
+		},
+		RoleRef: rbacv1.RoleRef{
+			APIGroup: rbacv1.GroupName,
+			Kind:     "Role",
+			Name:     name,
+		},
+	}
+}
+
+func dedupeStringSlice(s []string) []string {
```

Comment:
Think this can be replaced by `slices.Compact` since Go 1.22: https://pkg.go.dev/slices#Compact

Reply:
Ooh. Looks like that will look as long as the list is sorted first. And I am sorting lists before de-duping them, so that should work.